workflows4s-web-api-server/src/main/scala/workflows4s/web/api/service/RealWorkflowService.scala

```diff
@@ -82,15 +88,93 @@ class RealWorkflowService(
+          ProgressStep(
+            stepType = step.getClass.getSimpleName.replace("$", ""),
+            meta = ProgressStepMeta(
+              name = extractNameFromModel(step.toModel),
+              signalName = extractSignalNameFromModel(step.toModel),
+              operationName = extractOperationNameFromModel(step.toModel),
+              error = extractErrorFromModel(step.toModel),
+              description = extractDescriptionFromModel(step.toModel)
+            ),
+            result = step.result.map(res =>
+              ProgressStepResult(
+                status = if (res.value.isRight) "Completed" else "Failed",
+                index = res.index,
+                state = res.value.toOption.map(_.toString)
+              )
+            )
+          )
+        }.toList
+
+      case WIOExecutionProgress.HandleError(base, handler, meta, result) =>
+        convertProgressToSteps(base)
+
+      case single =>
+        List(ProgressStep(
+          stepType = single.getClass.getSimpleName.replace("$", ""),
+          meta = ProgressStepMeta(
+            name = extractNameFromModel(single.toModel),
+            signalName = extractSignalNameFromModel(single.toModel),
+            operationName = extractOperationNameFromModel(single.toModel),
+            error = extractErrorFromModel(single.toModel),
+            description = extractDescriptionFromModel(single.toModel)
+          ),
+          result = single.result.map(res =>
+            ProgressStepResult(
+              status = if (res.value.isRight) "Completed" else "Failed",
+              index = res.index,
+              state = res.value.toOption.map(_.toString)
+            )
+          )
+        ))
+    }
   }
+
+  private def extractNameFromModel(model: WIOModel): Option[String] = model match {
+    case WIOModel.RunIO(meta) => meta.name
+    case WIOModel.HandleSignal(meta) => meta.operationName
+    case WIOModel.Pure(meta) => meta.name
+    case WIOModel.Timer(meta) => meta.name
+    case WIOModel.Fork(_, meta) => meta.name
+    case _ => None
+  }
+
+  private def extractSignalNameFromModel(model: WIOModel): Option[String] = model match {
+    case WIOModel.HandleSignal(meta) => Some(meta.signalName)
+    case _ => None
+  }
+
+  private def extractOperationNameFromModel(model: WIOModel): Option[String] = model match {
+    case WIOModel.HandleSignal(meta) => meta.operationName
+    case _ => None
+  }
+
+  private def extractErrorFromModel(model: WIOModel): Option[String] = model match {
+    case WIOModel.RunIO(meta) => meta.error.map(_.name)
+    case WIOModel.HandleSignal(meta) => meta.error.map(_.name)
+    case WIOModel.Pure(meta) => meta.error.map(_.name)
+    case WIOModel.Dynamic(meta) => meta.error.map(_.name)
+    case _ => None
+  }
+
+  private def extractDescriptionFromModel(model: WIOModel): Option[String] = model match {
+    case WIOModel.RunIO(meta) => meta.description
+    case _ => None
+  }
```

Comment:
I find all those conversions a bit odd and atificial. What made you go this way? Teturning raw serialized `WIOExecutionProgress` should be enough for now I think. If we want to transform it into some more easily consumable form we should discuss it first.